### DISCUSSION TITLE
Occasionally getting connection refused error from Puma

### DISCUSSION THREAD

**Discussion by lijithv** [puma/puma#3203] [Q&A]
Puma version - 5.6.5
JRuby version - 9.1.17.0

We are experiencing intermittent connection refused errors in our JRuby app running on the Puma server. OS is Windows. These errors occur sporadically, particularly when the server load increases. Server resource usage is minimal. Only less than 30% CPU used and over 4 GB of free memory available. We have observed that the errors last only a few seconds (or less) before Puma resumes processing requests until the next occurrence.

Our Puma configuration consists of only the thread configuration set to 1:64. No errors are logged during these occurrences.

Any insights or suggestions on finding the cause would be highly appreciated.

**Comment by dentarg** [comment]
Have you tried latest Puma 6.x?

Have you tried to reproduce this locally? (Example app, same setup, adding load using wrk or some other tool (should be some examples in the repo or other issues / PRs))

**Reply by lijithv** [reply]
I couldn't replicate the issue locally with with similar setup. Unfortunately, this issue is happening at a customer env and upgrading puma is not an option for us.

**Comment by MSP-Greg** [comment]
Windows, JRuby, 1:64, particularly when the server load increases

I'll admit to not spending a lot of time testing JRuby on Windows.  IIRC, creating threads with JRuby does take more time that with MRI Rubies.

Also, I don't think I've ever tried to create an increasing load on a server with the tools I've used.  Or, the load is on or off.

So, if you have extra capacity in the server, you might try increasing the minimum thread count?

**Reply by lijithv** [reply]
Thank you for the suggestion. I raised the minimum thread count and even experimented with setting both the minimum and maximum to the same number. However, the problem still persists.

**Comment by xkkker** [comment]
Hi, did you fix the problem? We met the same issue with Puma 5.6.x.

**Reply by lijithv** [reply]
Hi @xkkker, unfortunately the issue is not resolved for us. For the time being, we have adjusted the UI logic to reduce the number of calls made to the server.

**Comment by rnhurt** [comment]
I think I'm seeing a similar error with AWS ECS and Rails.  We have a Rails application that is using the Curb gem to send requests to another Rails application.  Both are running the same stack (see below) and every once in a while we get a "Curl::Err::HostResolutionError" message.

- Puma: 6.4.2
- Ruby: 3.1.4
- Rails: 6.1.7.7
- Curb: 0.9.11
